packages/smart-contracts/src/contracts/ERC20EscrowToPayV1.sol

```diff
@@ -0,0 +1,385 @@
+        "lockDisputedFunds: Call to _withdraw failed!"
+        );
+
+        emit LockPeriodStarted(
+            _paymentRef,
+            disputeMapping[_paymentRef].tokentimelock,
+            disputeMapping[_paymentRef].paymentToken
+        );
+
+        return address(disputeMapping[_paymentRef].tokentimelock);
+    }
+
+    /**
+     * @notice Withdraw the locked funds from tokentimelock contract and transfer to payer.
+     * @param  _paymentRef Reference of the Invoice related.
+     */
+    function withdrawTimeLockedFunds(bytes memory _paymentRef) external OnlyPayer(_paymentRef) {
+        require(disputeMapping[_paymentRef].tokentimelock != TokenTimelock(0x0000000000000000000000000000000000000000) 
+            && disputeMapping[_paymentRef].amount != 0,
+            "ERC20EscrowToPay: No timelock contract found!"
+        );
+
+        TokenTimelock _tokentimelock = disputeMapping[_paymentRef].tokentimelock;
+        delete disputeMapping[_paymentRef];
+        
+        /// @dev Closes the tokentimelock and transfers the funds to the payer.
+        _tokentimelock.release();
+    
+        emit LockPeriodEnded(_paymentRef);
+    }
+    
+    /** 
+     * @notice Returns the endTime of the tokentimelock contract.   
+     * @param  _paymentRef Reference of the Invoice related.
+     */
+    function getLockPeriodEndTime(bytes memory _paymentRef) external view OnlyPayer(_paymentRef) returns (uint256 time) {
+        return disputeMapping[_paymentRef].tokentimelock.releaseTime();
+    }
+    
+    /**
+     * @notice Transfers paymentToken from payer to MyEscrow smartcontract.  
+     * @param _paymentRef Reference of the related Invoice.
+     * @dev Internal function to execute transferFrom() payer.
+     */
+    function _deposit(bytes memory _paymentRef) internal returns (bool result){
+        require(invoiceMapping[_paymentRef].paymentToken.transferFrom(
+            invoiceMapping[_paymentRef].payer,
+            address(this), 
+            (invoiceMapping[_paymentRef].amount + invoiceMapping[_paymentRef].feeAmount)), 
+            "_deposit: Can't _deposit tokens to Escrow, did you approve ERC20 token first?"
+        );
+        return true; 
+    }
+
+    /**
+     * @notice Withdraw the funds from the escrow.  
+     * @param _paymentRef Reference of the related Invoice.
+     * @dev Internal function to pay fees and withdraw funds to a given reciever.
+     */
+    function _withdraw(bytes memory _paymentRef, address _receiver) internal returns (bool result) {
+        /// Normal flow, funds are transfered to payee.
+        if (invoiceMapping[_paymentRef].amount != 0 ) {
+            uint256 _amount = invoiceMapping[_paymentRef].amount;
+            invoiceMapping[_paymentRef].amount = 0;
+            /// Give approval to transfer from ERC20EscrowToPayV1 => ERC20FeeProxy contract.
+            invoiceMapping[_paymentRef].paymentToken.approve(address(paymentProxy), _amount + invoiceMapping[_paymentRef].feeAmount);
```

Comment:
I'd be curious to know about the gas impacts but I assume it would be more efficient to approve the tokens we need "once and for all" with an administrtive permissionless task. You would do this approval only if needed.

Reply:
(same for the `approval` below as well.